deployer/src/deployer/search/__init__.py

```diff
@@ -0,0 +1,191 @@
+import json
+import re
+import time
+from pathlib import Path
+from collections import defaultdict
+
+import click
+from elasticsearch.helpers import streaming_bulk
+from elasticsearch_dsl.connections import connections
+from selectolax.parser import HTMLParser
+
+from .models import Document
+
+
+def index(
+    buildroot: Path,
+    url: str,
+    update=False,
+    no_progressbar=False,
+    priority_prefixes: (str) = (),
+):
+    # We can confidently use a single host here because we're not searching
+    # a cluster.
+    connections.create_connection(hosts=[url])
+    connection = connections.get_connection()
+    health = connection.cluster.health()
+    status = health["status"]
+    if status not in ("green", "yellow"):
+        raise click.ClickException(f"status {status} not green or yellow")
+
+    count_todo = 0
+    for file in walk(buildroot):
+        count_todo += 1
+
+    click.echo(f"Found {count_todo:,} (potential) documents to index")
+
+    # Confusingly, `._index` is actually not a private API.
+    # It's the documented way you're supposed to reach it.
+    index = Document._index
+    if not update:
+        click.echo(
+            "Deleting any possible existing index "
+            f"and creating a new one called {index._name}"
+        )
+        index.delete(ignore=404)
+        index.create()
+
+    search_prefixes = [None]
+    for prefix in priority_prefixes[::-1]:
+        search_prefixes.insert(0, prefix)
+
+    count_by_prefix = defaultdict(int)
+
+    already = set()
+
+    def generator():
+        for prefix in search_prefixes:
+            root = Path(buildroot)
+            if prefix:
+                root /= prefix
+            iterator = walk(root)
+            for doc in iterator:
+                if doc in already:
+                    continue
+                already.add(doc)
+                search_doc = to_search(doc)
+                if search_doc:
+                    count_by_prefix[prefix] += 1
+                    yield search_doc.to_dict(True)
+
+    def get_progressbar():
+        if no_progressbar:
+            return VoidProgressBar()
+        return click.progressbar(length=count_todo, label="Indexing", width=0)
+
+    count_done = 0
+    t0 = time.time()
+    with get_progressbar() as bar:
+        for x in streaming_bulk(connection, generator(), index=index._name):
+            count_done += 1
+            bar.update(1)
```

Comment:
It's not immediately clear to me why, but for some reason, when I ran the `search-index` command, the progress bar didn't display the final tick:
```sh
rjohnson-44864:deployer rjohnson$(maws_profile)$ poetry run deployer search-index ../client/build
The currently activated Python version 2.7.16 is not supported by the project (^3.7).
Trying to find and use a compatible version.
Using python3 (3.8.0)
Found 50,342 (potential) documents to index
Deleting any possible existing index and creating a new one called mdn_docs
Indexing  [#################################################################################-]   99%  00:00:00
Took 3m 5s to index 50,277 documents. Approximately 270.7 docs/second
```

Reply:
It's a fair comment. The reason it happens is that we count all the docs first. It's the simplest loop. It basically counts all the `index.json` files. But when it properly starts to iterate through the files, once the `index.json` has been opened, it gets a bit more picky. I.e. it skips certain files, in particular the pages whose `slug` endswith `/Index`. 
I'll see if there's something I can do about that.